### ISSUE TITLE
`redefined-outer-name` (`W0621`) - false positive on 3.12 type aliases

### ISSUE THREAD

**Issue by DetachHead** [pylint-dev/pylint#9884]
### Bug description

```python
type Foo[T] = T

def foo[T](): ...
```

### Command used

```shell
pylint a.py
```

### Pylint output

```python
W0621: Redefining name 'T' from outer scope (line 1) (redefined-outer-name)
```

### Expected behavior

no error

### Pylint version

```shell
pylint 3.2.6
astroid 3.2.4
Python 3.12.0 (tags/v3.12.0:0fb18b0, Oct  2 2023, 13:03:39) [MSC v.1935 64 bit (AMD64)]
```

**Comment by cdce8p** [comment]
It looks like the type parameter is bound to the wrong scope (module). This probably needs to be fixed in astroid.

**Comment by Pierre-Sassoulas** [comment]
Here's a snippet using overload from #10995 that I closed as duplicate :
```python
# PEP 695 type alias — `T` here is scoped to this statement only.
# Per PEP 695: "the type parameters are only visible within the scope
# of the statement."
type Alias[T] = T | list[T]


# This function's `T` is a different, locally-scoped type parameter.
# Python itself has no issue — this is valid, well-scoped code.
# pylint incorrectly warns:
#   W0621: Redefining name 'T' from outer scope (line 42)
def identity[T](x: T) -> T:
    return x


# Same issue with @overload — each overload introduces its own `T`.
@overload
def process(x: str) -> str: ...
@overload
def process[T](x: list[T]) -> T: ...  # W0621 (false positive)
def process[T](x: str | list[T]) -> str | T:  # W0621 (false positive)
    if isinstance(x, list):
        return x[0]
    return x
```
